mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+#![allow(dead_code)] // TODO remove when ready
+use std::{collections::HashSet, convert::TryInto, time::Duration};
+
+use async_trait::async_trait;
+use chrono::Utc;
+use futures_util::future::BoxFuture;
+use tokio::{net::TcpStream, stream::StreamExt};
+use tracing::{debug, error};
+
+use mqtt3::{
+    proto, Client, Event, IoSource, ShutdownError, SubscriptionUpdateEvent, UpdateSubscriptionError,
+};
+
+use crate::settings::Credentials;
+use crate::token_source::{SasTokenSource, TokenSource};
+
+const DEFAULT_TOKEN_DURATION_MINS: u64 = 60;
+const DEFAULT_TOKEN_DURATION_SECS: Duration = Duration::from_secs(60 * DEFAULT_TOKEN_DURATION_MINS);
+const DEFAULT_MAX_RECONNECT: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+pub struct ShutdownHandle(mqtt3::ShutdownHandle);
+
+impl ShutdownHandle {
+    pub async fn shutdown(&mut self) -> Result<(), ClientConnectError> {
+        self.0
+            .shutdown()
+            .await
+            .map_err(ClientConnectError::ShutdownClient)?;
+        Ok(())
+    }
+}
+
+pub struct TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    address: String,
+    token_source: Option<T>,
+}
+
+impl<T> IoSource for TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        let token_source = self.token_source.as_ref().cloned();
+
+        Box::pin(async move {
+            let expiry = Utc::now()
+                + chrono::Duration::from_std(DEFAULT_TOKEN_DURATION_SECS).unwrap_or_else(|_| {
+                    chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS.try_into().unwrap())
+                });
```

Comment:
unlike c# everything in Rust is an expression, so you can assign result of if-block to variable:
```rust
            let password = if let Some(ts) = token_source {
                match ts.get(&expiry).await {
                    Ok(x) => Some(x),
                    Err(e) => {
                        error!("Failed to get token for connection {} {}", address, e);
                        None
                    }
                }
            } else {
                None
            };
```

Reply:
Also there are some combinators to make conversion Result -> Option if you need. https://doc.rust-lang.org/std/result/enum.Result.html#method.ok